ui/src/app/resolvers/workspace.ts

```diff
@@ -0,0 +1,31 @@
+import {Injectable} from '@angular/core';
+import {ActivatedRouteSnapshot, Resolve} from '@angular/router';
+import {Observable} from 'rxjs/Observable';
+
+import {Workspace, WorkspaceAccessLevel, WorkspacesService} from 'generated';
+
+/**
+ * Flatten a layer of nesting
+ */
+interface WorkspaceData extends Workspace {
+  accessLevel: WorkspaceAccessLevel;
+}
+
+@Injectable()
+export class WorkspaceResolver implements Resolve<WorkspaceData> {
+  constructor(private api: WorkspacesService) {}
+
+  resolve(route: ActivatedRouteSnapshot): Observable<WorkspaceData> {
+    const ns: Workspace['namespace'] = route.params.ns;
+    const wsid: Workspace['id'] = route.params.wsid;
+
+    console.log(`Resolving Workspace ${ns}/${wsid}:`);
+    console.dir(route);
+
+    const call = this.api
+      .getWorkspace(ns, wsid)
+      .map(({workspace, accessLevel}) => ({...workspace, accessLevel}));
```

Comment:
Question: If I open up a workspace, and then am unshared from it, but continue clicking around, this will let me keep doing things as long as I don't refresh the page, correct?

Reply:
If you're unshared then the client doesn't know (unless we go some kind of SSE / websocket / long polling for permissions route) until the client attempts something that would require access, but that's true no matter how the workspace is loaded.  Its possible that the client could remain in an invalid state longer this way?  I'll have to ruminate on that.  At any rate, it should be the case that either way (route resolver or load in the component) you can keep doing things unless those things affect server state or reload client state from the server.